src/iop/useless.c

```diff
@@ -232,60 +290,185 @@ void cleanup_global(dt_iop_module_so_t *module)
+  // this only has to be provided if module settings or default_params need to depend on
+  // image type (raw?) or exif data.
 
-  self->widget = gtk_box_new(GTK_ORIENTATION_VERTICAL, DT_BAUHAUS_SPACE);
+  // can't be switched on for non-raw images:
+  module->hide_enable_button = !dt_image_is_raw(&module->dev->image_storage);
 
-  g->scale = dt_bauhaus_slider_new_with_range(self, 1, 100, 1, 50, 0);
-  dt_bauhaus_widget_set_label(g->scale, NULL, _("size"));
-  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->scale), TRUE, TRUE, 0);
-  g_signal_connect(G_OBJECT(g->scale), "value-changed", G_CALLBACK(scale_callback), self);
+  // if we are in darkroom, gui_init will already have been called and initialise module->gui_data
+  // it may then be necessary to also change the default values in widgets.
+  dt_iop_useless_gui_data_t *g = (dt_iop_useless_gui_data_t *)module->gui_data;
+  if(g)
+  {
+  }
+}
 
-  g->factor = dt_bauhaus_slider_new_with_range(self, 0.0, 1.0, 0.1, 0.5, 2);
-  dt_bauhaus_widget_set_label(g->factor, NULL, _("factor"));
-  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->factor), TRUE, TRUE, 0);
-  g_signal_connect(G_OBJECT(g->factor), "value-changed", G_CALLBACK(factor_callback), self);
+void gui_init(dt_iop_module_t *self)
+{
+  // allocates memory for the module's user interface in the darkroom and
+  // sets up the widgets in it.
+  // self->widget needs to be set to the top level widget.
+  // this can be a (vertical) box, a grid or even a notebook. modules that are
+  // disabled for certain types of pictures (for example non-raw) may use a stack
+  // where one of the pages contains just a label explaining why it is disabled.
+  // widgets that are directly linked to a field in params_t may be set up using the
+  // dt_bauhaus_..._from_params family. they take a string with the field
+  // name in the params_t struct definition. The $MIN, $MAX and $DEFAULT tags will be
+  // used to set up the widget (slider) ranges and default values and the $DESCRIPTION
+  // is used as the widget label. dt_bauhaus_an automatic callback is setup that updates the field
+  // whenever the widget is changed. additionally, gui_changed is called, if it exists,
+  // so that any other required changes, to dependent fields or to gui widgets, can be made.
+  // whenever self->params changes (switching images or history) the widget values have to
+  // be updated in gui_update.
+  // any widgets that are _not_ directly linked to a field need to have a custom callback
+  // function set up to respond to the "value-changed" signal.
+  // do not set the value of widgets or configure them depending on field values; this should
+  // be done in gui_update (or gui_changed or individual widget callbacks)
+  // if any default values for (slider) widgets or options (in comboboxes) depend on the
+  // type of picture, then the widgets have to be updated in reload_params.
```

Comment:
This is too long for me and needs some structure. Some of the comments here might work better with a worked example.

Reply:
This is probably the most crucial section as I expect gui_init is where people would look for information most, so I'm branching to all the other functions from here. I've broken it into paragraphs and moved one into the example section.

please try to read again!

Reply:
It's better like that. Still need to consistently start sentences with capital letters though. It really does help with readability (I try to make sure sentences don't start with the word 'darktable' or I can't sleep at night).

I still feel like I need to understand it all better in the first place to comment too much though.